pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go

```diff
@@ -959,97 +970,112 @@ func Test_ChooseUpgradePathInvariants(t *testing.T) {
 		}
 		return &clusterupgrade.Version{Version: curVersion}
 	}
+	// N.B. technically, v22.1.0 is the oldest version with release data, but we want to streamline input generation.
+	// By sticking with `minimumSupported`, we can avoid a special case where generated finalVersion has the same release
+	// series as `minimumSupported`.
+	oldestVersionSupported := minimumSupported
+	newestPossibleVersion := clusterupgrade.MustParseVersion("v25.1.0")
+	maxPossibleUpgrades, err := release.MajorReleasesBetween(&newestPossibleVersion.Version, &oldestVersionSupported.Version)
+	require.NoError(t, err)
 
 	generator := func(values []reflect.Value, rng *rand.Rand) {
 		// We don't always want to upgrade into the latest release (v25.1) so pick a release
-		// n predecessors older.
+		// n predecessors older, excluding oldestVersionSupported.
 		finalVersionOffset := rng.Intn(maxPossibleUpgrades)
 		finalVersion := randomNthPredecessor(rng, newestPossibleVersion, finalVersionOffset)
 		values[0] = reflect.ValueOf(finalVersion.Version.String())
 
-		maxUpgrades := rng.Intn(maxPossibleUpgrades-finalVersionOffset) + 1
-		upgradesRemaining := maxUpgrades - finalVersionOffset
+		upgradesRemaining := maxPossibleUpgrades - finalVersionOffset
+		require.Greater(t, upgradesRemaining, 0)
+		// Randomly choose minimumBootstrapVersion <= minimumSupportedVersion.
+		chooseMbvAndMsv := func() (*clusterupgrade.Version, *clusterupgrade.Version) {
```

Comment:
Simplified the logic of choosing both as well as allowing either to be "zero", i.e., unset.